### ISSUE TITLE
Compiling v17.10c with latest FairSoft fails on Fedora 28

### ISSUE THREAD

**Issue by PMunkes** [FairRootGroup/FairRoot#819]
This bug occurs on Fedora 28, with gcc 8.1.1 (the planned compiler of Ubuntu 18.10).
Compiling FairRoot v17.10c with the the oct17p1 release of FairSoft fails due to an incompatible version of XRootD. Trying to compile FairRoot v17.10c with the latest version of FairSoft (may18), which ships the fixed XRootd version, fails with the following output:

```
 In file included from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:9:
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h: In instantiation of ‘void BaseMQFileSink<InputPolicy, OutputPolicy>::Run() [with InputPolicy = fair::mq::policy::InputPolicy<RootDeserializer, TClonesArray, fair::mq::policy::PointerType, fair::mq::policy::OpNewCreator, fair::mq::policy::NullptrInitializer, fair::mq::policy::RawPtrDeleter>; OutputPolicy = RootOutFileManager<FairMBSRawItem>]’:
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h:48:18:   required from here
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h:56:17: error: request for member ‘Deserialize’ is ambiguous
                  Deserialize<typename InputPolicy::DeserializerType>(*msg, InputPolicy::fInput); // get data from message.
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 In file included from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:12:
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/base/MQ/policies/Serialization/RootSerializer.h:71:10: note: candidates are: ‘template<class T> void RootDeserializer::Deserialize(FairMQMessage&, std::unique_ptr<T>&)’
      void Deserialize(FairMQMessage& msg, std::unique_ptr<T>& output)
           ^~~~~~~~~~~
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/base/MQ/policies/Serialization/RootSerializer.h:63:10: note:                 ‘template<class T> void RootDeserializer::Deserialize(FairMQMessage&, T*&)’
      void Deserialize(FairMQMessage& msg, T*& output)
           ^~~~~~~~~~~
 In file included from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/PluginServices.h:13,
                  from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/Plugin.h:13,
                  from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/PluginManager.h:12,
                  from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/DeviceRunner.h:13,
                  from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/runFairMQDevice.h:9,
                  from /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:8:
 /opt/CBM/fairroot_src_v-17.10c-fairsoft_may18_root6/fairmq/FairMQDevice.h:83:10: note:                 'template<class Deserializer, class DataType, class ... Args> void FairMQDevice::Deserialize(FairMQMessage&, DataType&&, Args&& ...) const’
      void Deserialize(FairMQMessage& msg, DataType&& data, Args&&... args) const
           ^~~~~~~~~~~
 make[2]: *** [examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/build.make:63: examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/run/runMBSSink.cxx.o] Error 1
 make[1]: *** [CMakeFiles/Makefile2:6122: examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/all] Error 2
 make[1]: *** Waiting for unfinished jobs....
```

Presumably this happens due to changes in FairMQ, although I am not familiar enough with the codebase in question. 

I am currently testing a modified version of FairSoft oct17p1, in which I included the XRootd version 4.8.3, and I am considering to submit a patch with this change for FairSoft oct17p2, but fixing this in FairRoot might be a better idea.

**Comment by dennisklein** [comment]
The dev branch of FairRoot fixes this problem with gcc 8, see aeda09a. I guess, we can backport the fix tomorrow to the latest release.

**Comment by PMunkes** [comment]
Backporting commit [aeda09a](https://github.com/FairRootGroup/FairRoot/commit/aeda09aa8450ca1ce4dd9d8663648e97e0677ecf) to 17.10c fixes the error and FairRoot can be compiled without errors. A few tests fail, the log is attached.
`The following tests FAILED:
	 40 - run_reco_timebased_TGeant4 (Timeout)
	 41 - MQ.ex7-parameters (Timeout)
	 42 - MQ.ex-serialization1 (Timeout)
	 43 - MQ.ex-serialization2 (Timeout)
	 47 - ex9_static (Timeout)
	 48 - ex9a_static (Timeout)
Errors while running CTest`
[LastTest.log](https://github.com/FairRootGroup/FairRoot/files/2378610/LastTest.log)
The first two of these fail with a message close to this:
`Error in <TList::Clear>: A list is accessing an object (0x7fb16875d880) already deleted (list name = TList)`

**Comment by dennisklein** [comment]
FairRoot [17.10d](https://github.com/FairRootGroup/FairRoot/releases/tag/v-17.10d) should build with GCC 8.

I can confirm the timeouts of some of the tests with the same `TList::Clear` double free error, but for only 4 tests failed:

```
The following tests FAILED:
	 40 - run_reco_timebased_TGeant4 (Timeout)
	 42 - MQ.ex-serialization1 (Timeout)
	 43 - MQ.ex-serialization2 (Timeout)
	 47 - ex9_static (Timeout)
```

@PMunkes Do you see the double `TList::Clear` error also in your production?

@fuhlig1 @MohammadAlTurany Was FairRoot `17.10` + FairSoft `may18` in your test matrix? Was it working at some point in time? Do we fix it?

**Comment by PMunkes** [comment]
FairRoot [v-17.10d](https://github.com/FairRootGroup/FairRoot/releases/tag/v-17.10d) does indeed build without modifications on GCC 8.1.1.

However, I do see these `TList::Clear` double free errors and 5 tests fail:

> The following tests FAILED:
> 	  1 - FairMQ.Protocols (Timeout)
> 	 40 - run_reco_timebased_TGeant4 (Timeout)
> 	 42 - MQ.ex-serialization1 (Timeout)
> 	 43 - MQ.ex-serialization2 (Timeout)
> 	 48 - ex9a_static (Timeout)
Interestingly the failing tests have changed from the modified 17.10c with the backported commit aeda09a.

I think I can save you a little bit of time investigating test 40, as the error is already occuring during simulation. On my machine, for some reason during event 90 Geant4(?) generates two Timestamps of approximately 7.2e11. I have been unable to further debug this, as root is quite effective in preventing the use of gdb.

I have attached the Full Log:
[LastTest.log](https://github.com/FairRootGroup/FairRoot/files/2412977/LastTest.log)

On a machine with Ubuntu 16.04 with GCC 5.4.0 I get 2 failed tests:
> 	 42 - MQ.ex-serialization1 (Timeout)
> 	 43 - MQ.ex-serialization2 (Timeout)

**Comment by fuhlig1** [comment]
Hi @dennisklein,

> @fuhlig1 @MohammadAlTurany Was FairRoot 17.10 + FairSoft may18 in your test matrix? Was it working at some point in time? Do we fix it?

we never test this mixed version. We only test FairRoot and FairSoft versions which were released together. Such mixed setups normally work but there is no guarantee from our side. In this special case I can imagine problems sind FairSoft may18 and FairRoot v17.10 both contain FairMq and FairLogger.
 I would propose either to use  

   FairRoot v-17.10 together with FairSoft oct17p1

or

   FairRoot v18 together with FairSoft may18

Ciao

Florian

**Comment by PMunkes** [comment]
Hi, 
Unfortunately Fairsoft oct17p1 does not work, since the included Xrootd-Version doesn't compile with GCC 8.1. Also there is no current official release of FairRoot v18, so I used the two latest releases together. 
Building FairRoot v18.0.1 with FairSoft may18 fails due to the ambiguos method call from the opening post:

> In file included from /opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:9:
/opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h: In instantiation of ‘void BaseMQFileSink<InputPolicy, OutputPolicy>::Run() [with InputPolicy = fair::mq::policy::InputPolicy<RootDeserializer, TClonesArray, fair::mq::policy::PointerType, fair::mq::policy::OpNewCreator, fair::mq::policy::NullptrInitializer, fair::mq::policy::RawPtrDeleter>; OutputPolicy = RootOutFileManager<FairMBSRawItem>]’:
/opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h:48:18:   required from here
/opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/base/MQ/devices/BaseMQFileSink.h:56:17: error: request for member ‘Deserialize’ is ambiguous
                 Deserialize<typename InputPolicy::DeserializerType>(*msg, InputPolicy::fInput); // get data from message.
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:12:
/opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/base/MQ/policies/Serialization/RootSerializer.h:71:10: note: candidates are: ‘template<class T> void RootDeserializer::Deserialize(FairMQMessage&, std::unique_ptr<T>&)’
     void Deserialize(FairMQMessage& msg, std::unique_ptr<T>& output)
          ^~~~~~~~~~~
/opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/base/MQ/policies/Serialization/RootSerializer.h:63:10: note:                 ‘template<class T> void RootDeserializer::Deserialize(FairMQMessage&, T*&)’
     void Deserialize(FairMQMessage& msg, T*& output)
          ^~~~~~~~~~~
In file included from /opt/CBM/fairsoft_may18_root6/installation/include/fairmq/PluginServices.h:13,
                 from /opt/CBM/fairsoft_may18_root6/installation/include/fairmq/Plugin.h:13,
                 from /opt/CBM/fairsoft_may18_root6/installation/include/fairmq/PluginManager.h:12,
                 from /opt/CBM/fairsoft_may18_root6/installation/include/fairmq/DeviceRunner.h:13,
                 from /opt/CBM/fairsoft_may18_root6/installation/include/fairmq/runFairMQDevice.h:9,
                 from /opt/CBM/fairroot_src_v18.0.1-fairsoft_may18_root6/examples/MQ/Lmd/run/runMBSSink.cxx:8:
/opt/CBM/fairsoft_may18_root6/installation/include/fairmq/FairMQDevice.h:83:10: note:                 ‘template<class Deserializer, class DataType, class ... Args> void FairMQDevice::Deserialize(FairMQMessage&, DataType&&, Args&& ...) const’
     void Deserialize(FairMQMessage& msg, DataType&& data, Args&&... args) const
          ^~~~~~~~~~~
make[2]: *** [examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/build.make:63: examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/run/runMBSSink.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:3939: examples/MQ/Lmd/CMakeFiles/ex-lmd-mbs-sink.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

Backporting aeda09a once again fixes the build issues, however two tests still fail:
> The following tests FAILED:
	 24 - MQ.ex-serialization1 (Timeout)
	 25 - MQ.ex-serialization2 (Timeout)
Errors while running CTest

The Full Log can be found [here](https://github.com/FairRootGroup/FairRoot/files/2414573/LastTest.log). 


This is going to affect more people going forward, as Ubuntu 18.10 will also ship with this compiler, so we either need a new release of FairRoot or a clear statement which tag to use in this situation.
Another solution, one that would prevent situations like this going forward, would be to migrate all downstream projects to aliBuild, where the compiler version would be under the control of the maintainer.

**Comment by dennisklein** [comment]
> Also there is no current official release of FairRoot v18, so I used the two latest releases together.

There is not? How do you conclude that?

Nevertheless, we will definitely support GCC 8. If you are fine with using the most recent releases of FairRoot and FairSoft, I will focus on fixing it there.

**Comment by PMunkes** [comment]
> >Also there is no current official release of FairRoot v18, so I used the two latest releases together.
> 
> There is not? How do you conclude that?
> 

I looked at the releases page: [releases](https://github.com/FairRootGroup/FairRoot/releases). The latest release of RairRoot on there was v 17.10c at the time of this post.

As far as I am concerned v18.0.1, with the backported fix, should be sufficient for needs. I will update you later if CbmRoot works with it and close this issue if it does.

**Comment by dennisklein** [comment]
Ill prepare the backport for the gcc 8 fix for you and release it as 18.0.2 today. Ill also discuss in the next meeting on how to address the failing tests. And Ill clear up if release 18.0 was meant to be inofficial, because from what I remember it was a normal release, we just changed the versioning schema. But well, I could be wrong.

**Comment by dennisklein** [comment]
Status update:

| **FairRoot** | **FairSoft** | **Compiler** | **Failed tests** |
| --- | --- | --- | --- |
| `v-17.10d` | `may18` | GCC 8.1.1 | run_reco_timebased_TGeant4 (Timeout), MQ.ex-serialization1 (Timeout), MQ.ex-serialization2 (Timeout), ex9_static (Timeout) |
| `v18.0.2` | `may18` | GCC 8.1.1 | MQ.ex-serialization1 (Timeout), MQ.ex-serialization2 (Timeout) |
| `v18.0.2` | `may18` | GCC 4.9.2 | MQ.ex-serialization1 (Timeout) |
| `dev` (`acbd45f9d`) | `may18` | GCC 8.1.1 | run_sg (Failed), MQ.ex-serialization1 (Timeout), MQ.ex-serialization2 (Timeout) |

@PMunkes sees almost the same failing test, sometimes one or the other more. The timeouts seem to be caused by a `TList::Clear` trying to double free a list entry. Sometimes they show up as segfault only without error message about the double `TList::Clear`. So we are dealing with a race and an object ownership problem which remain unresolved.

*Clarification*: FairRoot releases `v18.0.x` are just regular official releases. We apparently forgot to post the release notes, which will be added soonish.

**Comment by dennisklein** [comment]
Status update:

| **FairRoot** | **FairSoft** | **Compiler** | **Failed tests** |
| --- | --- | --- | --- |
| `v18.0.3` | `may18` | GCC 4.9.2 | _none_ |
| `v18.0.3` | `may18` | GCC 8.1.1 | _none_ |
| `dev` (`dd228beef`) | `may18` | GCC 8.1.1 | _none_ |

I will probably not fix the failing examples on the 17.10 release any more unless anyone is really blocked by it.

@PMunkes Thank you for helping to improve FairRoot! We believe your issue is resolved in the recent releases `17.10d` and `v18.0.3` of FairRoot (tested w/ FairSoft `may18`). Please reopen or create a new issue, if we missed anything.

**Comment by PMunkes** [comment]
Thanks Dennis, I can actually add another datapoint to this, as GCC 8.2.1 is also running perfectly on FairRoot v18.0.3. I will try version v-17.10d after this build. Thank you for your work in fixing this.

**Comment by PMunkes** [comment]
As an addendum to my last post: v17-10d does not work with GCC 8.2.1, but I feel that my issue has been fully resolved with v18.0.3.
Thank all of you for your hard work.

**Comment by dennisklein** [comment]
> v-17.10d does not work with GCC 8.2.1

Does it not compile? Or is it just some failing test?

**Comment by PMunkes** [comment]
There is a list of about 6 failed tests. Unfortunately I don't currently have access to my other machine, so I can't post it here. I will run a new build once I get home, but I am currently stuck at university for a few more hours.

**Comment by dennisklein** [comment]
As long as the compilation works, I think we are good (I consider the failing tests as a minor issue here). As @fuhlig1 pointed out, the combination of FairRoot `17.10` and FairSoft `may18` was never properly tested, even with older compilers.

As a side note: Our tests still have execution robustness issues. I can e.g. produce failures by running them in parallel on the same machine for example. I believe also, that some tests will still occasionally fail because of bugs, but those are probably minor.
